### ISSUE TITLE
ping

### ISSUE THREAD

**Issue by splitice** [nodejs/quic#37]
An interface to send QUIC ping frames would be really handy for long lived connections. It's also common in the existing QUIC prototype libraries that exist for javascript.

**Comment by jasnell** [comment]
We certainly could, but handling the ping responses are a bit problematic in that they come in as normal Acks that are not surfaced via the ngtcp2 API. This may need to wait until ngtcp2 provides the necessary API bits. That said, ngtcp2 does keep rolling RTT stats that we do have access to.

**Comment by splitice** [comment]
Yes looking the provided API there could be some complexity here. However it's reasonably common.

For context we are experimenting with porting an existing protocol implemented on a basic pure-js QUIC implementation (without TLS) to this implementation. This was a feature provided by that implementation, however for our usage we could easily do it at the stream layer.

**Comment by jasnell** [comment]
So, investigating further on this, there are ways we can *send* a ping but ngtcp2 does not provide any mechanism for knowing that a ping frame has been acknowledged. Before going much further on this, we should agree on what the actual use case need here is.. specifically is the ping intended to:

1. Detect whether the connection is still active
2. Keep the connection active to avoid timeouts
3. Measure the RTT latency for the connection

For 1, with QUIC, a connection is only active when there's data actually being transmitted. There may be periods of dormancy that are limited by the idle_timeout on configuration. We can easily report on the gaps of time between activity as we are actively tracking that already (e.g. "there's been no activity in the past 20 milliseconds"). We don't need PING support for this.

For 2, we can send PING frames in a fire-and-forget type of way with a hope that it'll keep the connection open but there's really no guarantee. If we don't need guarantees, then just having the API to send a PING is sufficient.

For 3, ngtcp2 reports rolling RTT statistics that are updated every time frames are processed. We don't have to send a PING to tap into these metrics. We can simply surface them on every QuicSession instance for applications to tap into any time they want.

Are there any other use cases that we may need to cover?

**Comment by splitice** [comment]
Yes that describes the three reasons we send pings. Right now to get our port of that application working we ended up with a quic stream kept open doing 1 byte req & responses to imitate that behavior (not efficient, but workable).

**Comment by jasnell** [comment]
That's not an unreasonable approach to be honest, and it's something that we could make more efficient by implementing at the native layer.
